### PR TITLE
Allow changing Node names since Node now have IDs

### PR DESCRIPTION
**Pull request by pierresouchay**
This is a proposal for being able to rename nodes when their IP does not change.

It should fix https://github.com/hashicorp/consul/issues/3974

### CODE REVIEW

**Comment by pierresouchay** [comment]
I basically removed the IP address limitation since node ID is supposed to be immutable, furthermore it is very likely that a nodes does change its IPs as well as its name at the same time for operation reasons (changing node's domain for instance)

**Comment by kamaradclimber** [comment]
Any news on that PR?

File: agent/consul/state/catalog.go [hunk 1]
```diff
@@ -355,9 +355,14 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
 		if existing != nil {
 			n = existing.(*structs.Node)
 			if n.Node != node.Node {
-				return fmt.Errorf("node ID %q for node %q aliases existing node %q",
-					node.ID, node.Node, n.Node)
+				// We are actually renaming a node, remove its reference first
+				err := s.deleteNodeTxn(tx, idx, n.Node)
```
Comment (mkeeler) [inline agent/consul/state/catalog.go]: @pierresouchay One concern we have is that this will allow multiple nodes to use the same name with different UUIDs. This might break some things and at the very least could be confusing with regards to metrics and labels. 

I think this could be remedied by refactoring this block of code and the lookup by name right after this. Essentially we would want to only delete the node when it has an id, the name is different **and** the name is not already associated with another node.

Comment (pierresouchay) [reply agent/consul/state/catalog.go]: @mkeeler Great, I made the fix, I also fix another issue regarding case insensitivity that lead us to a major downtime of Consul.

Since DNS is case insensitive and DB as issues when similar names with different cases are added, check for unicity based on case insensitivity.

Following another big incident we had in our cluster, we also validate that adding/renaming a not does not conflicts with case insensitive matches.

We had the following error once:

* one node called: mymachine.MYDC.mydomain was shut off
* another node (different ID) was added with name: mymachine.mydc.mydomain before 72 hours

When restarting the consul server of domain, the consul server restarted failed to start since it detected an issue in RAFT database because mymachine.MYDC.mydomain and mymachine.mydc.mydomain had the same names.

Checking at registration time with case insensitivity should definitly fix those issues and avoid Consul DB corruption.

**Comment by pierresouchay** [comment]
> Just the one concern. If you can address that then I think we can merge this.

@mkeeler DONE in last commit (and improved, see commit message)

**Comment by banks** [comment]
@pierresouchay We had to revert this as we found a bunch of edge cases and test failures caused by it.

In debugging them, we realised that this is actually a breaking change to the API even if we did fix the edge cases which means we can't merge it until next major release (1.3) anyway.

We'll respond with more details in a bit, just in middle of getting 1.2.1 ready.

**Comment by mkeeler** [comment]
@pierresouchay Now for the full details.

We found a bunch of tests that were failing that went something like this:

1. Create a test agent (which generates a uuid for the node)
2. Issue a Catalog.Register request without the UUID (this removes the UUID from the node)
3. Later on do an anti-entropy state sync which includes the nodes UUID in an internal Catalog.Register RPC call.

Step 3 there was failing due to the else block you added (at our request unfortunately) that was ensuring we didn't add a node with an id when that node name was already taken.

The thought was that other users could be using Consul in a way that could be similar to our tests with registering services for a node that consul is running on and flipping things back and forth between having an id and then not having one (and eventually having AE syncing issues).

After more investigation current Consul behavior allows two things that we didn't want to change in a patch release:

1. Removing a UUID from a node. This really shouldn't be allowed but it may be relied upon by other users accidentally (just like our tests)
2. Adding a UUID to a node without one.

The PR had removed the ability to add a UUID to a node without one. While I think this is correct behavior, it probably needed to wait until a 1.3 release.

What I think we should do in the short term (1.2.2 probably) is to reduce the scope of this PR slightly. In order to preserve the behavior of being able to add a uuid to a node I think all that has to be done is to just remove the else block where there is no existing node with the new nodes uuid. Then longer term we can introduce other changes to prevent adding a uuid to a node without one (maybe optional unless an option is specified) and to prevent removing uuids from a node.

**Comment by pierresouchay** [comment]
@mkeeler ok, I think I understand.

What about having the same exact patch, but with an option such as `allow_node_renaming` defaulting to false for now (basically keeping old behaviour - maybe defaulting to true in later versions), and having the same exact behavior as my PR when set to true?

We had several incidents related to renaming:
 - several times, some minor ones (only nodes where impacted, but when those are your main DNS servers, it is not funny)
 - One very major as described in the commit message or node renaming, basically, one occurence of https://github.com/hashicorp/consul/issues/3217 ( => which forces us to restart each Consul server sequentially to fix it - the only workaround known to work in our experience) that did cause a database corruption on Consul servers (basically, upper/lower case in naming of 2 nodes alive at the same time) => we were unable to respawn our cluster server for several very long minutes (the leader was down, the other ones did not want to take lead due to DB corruption)

Pierre Souchay

**Comment by banks** [comment]
@pierresouchay I think the behaviour for the cases where either the existing or the registering node have no ID are pretty debatable and need more thought regardless of the fact they are a breaking change. And I think your issue is serious enough that it should be prevented.

The way I understand it, your patch only needs a slight change to allow it to prevent all the conditions you saw in production (assuming all your clients have IDs) without changing any of the questionable edge cases around nodes in state store or registration requests without IDs. I don't think it's a breaking change to prevent allow rename where IDs are present in both request and state store node with matching name and that seems to fix your actual cases right?

So if you made those changes to the PR (and added tests to verify that the behaviour when either incoming or existing node has a blank ID remains unchanged) I think we get a PR that fixes all your legit issues for everyone, and is not backward incompatible so can be released before 1.3.

Or do I misunderstand?

**Comment by pierresouchay** [comment]
@banks Ok, I am gonna have a try to update a review with the option then

**Comment by mkeeler** [comment]
@banks I think you and I are on the same page

Allowing to rename a node when the ID matches AND the name isn't used elsewhere is safe and is something we can put in today for the next patch release.

The bit that the PR added extra that was problematic was that it preventing updating the node registration with an ID if it had been registered without an ID.


Removing this hunk of code should be sufficient to get rid of any compatibility issues: https://github.com/pierresouchay/consul/blob/fecae3de21accbde40ded20a1a05be25b9e27787/agent/consul/state/catalog.go#L386-L391

In the longer term I think we are going to need to have a discussion about how we want to handle this. The safest approach is to add in that hunk that disallows adding an ID to a node that doesn't have one but where the names match and additionally prevent nodes with ids from being updated to not have them. This is going to require lots of updates to our tests (probably all of them that use a TestAgent) and would also disallow upgrading from nodes without ids to nodes with ids (without doing a leave and rejoin). Anyways that is a topic for later discussion. To just get the ability to rename nodes with ids should be simply to remove that hunk of code mentioned above. I did so locally and all the tests still pass (including the new ones).
